vm/src/stdlib/itertools.rs

```diff
@@ -65,37 +63,45 @@ mod decl {
         fn next(zelf: &Py<Self>, vm: &VirtualMachine) -> PyResult<PyIterReturn> {
-            loop {
-                let pos = zelf.cur_idx.load();
-                if pos >= zelf.iterables.len() {
-                    break;
-                }
-                let cur_iter = if zelf.cached_iter.read().is_none() {
-                    // We need to call "get_iter" outside of the lock.
-                    let iter = zelf.iterables[pos].clone().get_iter(vm)?;
-                    *zelf.cached_iter.write() = Some(iter.clone());
-                    iter
-                } else if let Some(cached_iter) = (*zelf.cached_iter.read()).clone() {
-                    cached_iter
-                } else {
-                    // Someone changed cached iter to None since we checked.
-                    continue;
-                };
-
-                // We need to call "next" outside of the lock.
-                match cur_iter.next(vm) {
-                    Ok(PyIterReturn::Return(ok)) => return Ok(PyIterReturn::Return(ok)),
-                    Ok(PyIterReturn::StopIteration(_)) => {
-                        zelf.cur_idx.fetch_add(1);
-                        *zelf.cached_iter.write() = None;
-                    }
-                    Err(err) => {
-                        return Err(err);
+            let next = || {
+                let source = zelf.source.read().clone();
+                match source {
+                    None => {
+                        return Ok(PyIterReturn::StopIteration(None));
                     }
+                    Some(source) => loop {
+                        let active = zelf.active.read().clone();
+                        match active {
+                            None => match source.next(vm) {
+                                Ok(PyIterReturn::Return(ok)) => {
+                                    *zelf.active.write() = Some(ok.get_iter(vm)?);
+                                }
+                                Ok(PyIterReturn::StopIteration(_)) => {
+                                    return Ok(PyIterReturn::StopIteration(None));
+                                }
+                                Err(err) => {
+                                    return Err(err);
+                                }
+                            },
+                            Some(active) => match active.next(vm) {
+                                Ok(PyIterReturn::Return(ok)) => {
+                                    return Ok(PyIterReturn::Return(ok));
+                                }
+                                Ok(PyIterReturn::StopIteration(_)) => {
+                                    *zelf.active.write() = None;
+                                }
+                                Err(err) => {
+                                    return Err(err);
+                                }
+                            },
+                        }
+                    },
                 }
```

Comment:
📎 "_It looks like you're `match`ing variants of an `Option`. Would you like help?_" 📎

```rust
if let Some(source) = zelf.source.read().clone() {
    loop {
        if let Some(active) = zelf.active.read().clone() {
            match active.next(vm) {
                Ok(PyIterReturn::Return(ok)) => {
                    return Ok(PyIterReturn::Return(ok));
                }
                Ok(PyIterReturn::StopIteration(_)) => {
                    *zelf.active.write() = None;
                }
                Err(err) => {
                    return Err(err);
                }
            }
        } else {
            match source.next(vm) {
                Ok(PyIterReturn::Return(ok)) => {
                    *zelf.active.write() = Some(ok.get_iter(vm)?);
                }
                Ok(PyIterReturn::StopIteration(_)) => {
                    return Ok(PyIterReturn::StopIteration(None));
                }
                Err(err) => {
                    return Err(err);
                }
            }
        }
    }
} else {
    Ok(PyIterReturn::StopIteration(None))
}
```